### ISSUE TITLE
MSSQL not connecting after 7.1.x

### ISSUE THREAD

**Issue by Alicmgz** [grafana/grafana#27224]
After version 7.1.x mssql, can not be connected to MSSQL databases with same setting in previous versions.

**Comment by xlson** [comment]
Hi,

Can you please update or recreate your issue using the [bug report template](https://github.com/grafana/grafana/issues/new?labels=type%3A+bug&template=1-bug_report.md) and add information about what version of MSSQL you are using as well as any other information that might be specific to your environment. Thank you.

**Comment by Drewster727** [comment]
same issue here. I had no issues on 7.1.3 then upgraded to 7.1.5 and I'm seeing this issue.

Data Sources "Saved & Test":
![image](https://user-images.githubusercontent.com/4528753/91329726-1e6c2e80-e78e-11ea-951b-507c67985848.png)

Dashboard that uses annotations sourced from SQL:
![image](https://user-images.githubusercontent.com/4528753/91329734-21ffb580-e78e-11ea-8fdc-26e76bd57485.png)

**Comment by Drewster727** [comment]
I rolled back to 7.1.3 and the problem went away. Not sure if this was introduced in 7.1.4 or 7.1.5

**Comment by xlson** [comment]
@Drewster727 it would be very helpful if you could add a bit more detail. What version of MSSQL are you running? Is the connection encrypted? Also, can you please configure debug logging fort mssql and upload the grafana log of the connection attempt?

```
filters = tsdb.mssql:debug
```

**Comment by kylebrandt** [comment]
Maybe something as a result of https://github.com/grafana/grafana/pull/27012

https://github.com/grafana/grafana/blob/82a235b54c4bf04a52952fe22a020db882303b8c/go.mod#L5-L7

**Comment by Drewster727** [comment]
> @Drewster727 it would be very helpful if you could add a bit more detail. What version of MSSQL are you running? Is the connection encrypted? Also, can you please configure debug logging fort mssql and upload the grafana log of the connection attempt?
> 
> ```
> filters = tsdb.mssql:debug
> ```

@xlson Just spun up another copy of 7.1.5 and tested again. Some details about my setup and connections to MS SQL:
- Running inside of a docker container
- Not using encryption
- Using a domain account to connect (it has access)
- Not specifying a port, but 1433 is correct (default)

![image](https://user-images.githubusercontent.com/4528753/91889867-2c212880-ec54-11ea-9982-dc4bcbf8f3ac.png)

I get the following when saving:
```
Unable to open tcp connection with host 'SERVERINTSQL01:1433': dial tcp 10.10.1.17:1433: connect: connection refused
```

Here's the debug logs (not very useful?):
```
t=2020-09-01T18:14:13+0000 lvl=dbug msg="Received command to update data source" logger=datasources url=SERVERINTSQL01\\INTERNAL
t=2020-09-01T18:14:13+0000 lvl=dbug msg="Parsing MSSQL URL" logger=tsdb.mssql url=SERVERINTSQL01\\INTERNAL
t=2020-09-01T18:14:13+0000 lvl=dbug msg="Recognized as ODBC URL format" logger=tsdb.mssql url=SERVERINTSQL01\\INTERNAL
t=2020-09-01T18:14:13+0000 lvl=dbug msg="Parsing MSSQL URL" logger=tsdb.mssql url=SERVERINTSQL01\\INTERNAL
t=2020-09-01T18:14:13+0000 lvl=dbug msg="Recognized as ODBC URL format" logger=tsdb.mssql url=SERVERINTSQL01\\INTERNAL
t=2020-09-01T18:14:13+0000 lvl=dbug msg="Generating connection string" logger=tsdb.mssql url=SERVERINTSQL01\\INTERNAL host=SERVERINTSQL01\\INTERNAL port=1433
t=2020-09-01T18:14:13+0000 lvl=info msg="Request Completed" logger=context userId=1 orgId=1 uname=admin method=POST path=/api/tsdb/query status=400 remote_addr=10.60.108.175 time_ms=33 size=371 referer=http://grafana01:3001/datasources/edit/1/
```

Last but not least, I tested with 7.1.4 and it shows the same issue. 7.1.3 works fine for me, so the issue definitely starts with 7.1.4.

Thanks guys!

**Comment by marefr** [comment]
> Maybe something as a result of #27012
> 
> https://github.com/grafana/grafana/blob/82a235b54c4bf04a52952fe22a020db882303b8c/go.mod#L5-L7

It was acutally #26188 introducing these changes and it will be included in Grafana v7.2. Seems like maybe a cherry pick for v7.1.5 didn't go as expected. However, not sure that's the reason.

Maybe related to https://github.com/denisenkom/go-mssqldb/pull/543. 

@Alicmgz @Drewster727 Could you please try changing your host setting to `:<server>\<instance>:0` to see if that makes any difference?

**Comment by marefr** [comment]
For reference some of these changes have most likely introduced this regression: https://github.com/denisenkom/go-mssqldb/compare/2be1aa5...b91950f

**Comment by BloodShura** [comment]
This also affects PostgreSQL datasource.

**Comment by marefr** [comment]
@BloodShura what? how? Please include detailed information. Thanks

**Comment by baur** [comment]
Same problem with 
grafana-7.1.5 & grafana-7.2.0-beta2
![image](https://user-images.githubusercontent.com/2898361/93865214-d803e400-fce7-11ea-8b35-1736705ec165.png)

**Comment by marefr** [comment]
@baur thanks. Have your tried `<server>\<instance>:0` or `<instance name>:0` as host?

**Comment by baur** [comment]
@marefr yes, <server>\<instance>:0 resolves problem

**Comment by marefr** [comment]
@baur thanks a lot :+1: I opened a bug report with the mssql library: https://github.com/denisenkom/go-mssqldb/issues/605

**Comment by Mario-Hofstaetter** [comment]
The Port `:0` workaround saved me too.

@marefr Their was feedback in https://github.com/denisenkom/go-mssqldb/issues/605#issuecomment-720879497 (could not reproduce ...), could you please check with the author feedback?

Also, there may have been a new release meanwhile, grafana uses this version?
https://github.com/grafana/grafana/blob/master/go.mod#L25 
which indicates `20200910202707` so 2020-09-10 ?

There has been a "first release" `0.9.0` https://github.com/denisenkom/go-mssqldb/releases 
on 2020-11-03, could grafana upgrade to it?

Are there other, more mature MSSQL drivers available in go? Reading https://github.com/denisenkom/go-mssqldb/issues/605#issuecomment-713549417 (_"It's not really production-ready"_) made me a little uncomfortable 👀

**Comment by Mario-Hofstaetter** [comment]
Still not working without port `:0` in grafana `7.5.1` containing https://github.com/grafana/grafana/pull/32347

**Comment by aknuds1** [comment]
@Mario-Hofstaetter this bug is unrelated to #32347. I have what I think is a fix in #32417.

**Comment by aknuds1** [comment]
Can those affected by this bug test my PR #32417?
